File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.js
/**
 * @param {string[]} words
 * @param {string} chars
 * @return {number}
 */
var countCharacters = function(words, chars) {
    const charCount = new Map();
    let result = 0;
    
    for (let char of chars){
        charCount.set(char, (charCount.get(char) || 0) + 1);
    }
    
    for (let word of words){
        if (formWord(word, charCount)){
            result += word.length;
        }
    }
    
    return result;
};

function formWord(word, charCount){
    const wordCount = new Map();
    const wordLength = word.length;
    
    for (let i = 0; i < wordLength; i++){
        const currentCount = (wordCount.get(word[i]) || 0) + 1;
        wordCount.set(word[i], currentCount);
        
        if (!charCount.has(word[i]) || currentCount > charCount.get(word[i])){
            return false;
        }
    }
    
    return true;
}

File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.py
class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        char_count = Counter(chars)
        result = 0
        
        for word in words:
            word_count = {}
            word_length = len(word)
            for i, c in enumerate(word):
                if c in word_count:
                    word_count[c] += 1
                else:
                    word_count[c] = 1
                
                if c not in char_count or word_count[c] > char_count[c]:
                    break
                
                if i == word_length - 1:
                    result += word_length
        
        return result
                    
                
            